File: Python_Books/算法图解/grokking_algorithms_COPY/06_breadth-first_search/javascript/01_breadth-first_search.js
'use strict';

function person_is_seller(name) {
  return name[name.length-1] === 'm';
}

const graph = {};
graph["you"] = ["alice", "bob", "claire"];
graph["bob"] = ["anuj", "peggy"];
graph["alice"] = ["peggy"];
graph["claire"] = ["thom", "jonny"];
graph["anuj"] = [];
graph["peggy"] = [];
graph["thom"] = [];
graph["jonny"] = [];


function search(name) {
  let search_queue = [];
  search_queue = search_queue.concat(graph[name]);
  // This array is how you keep track of which people you've searched before.
  const searched = [];
  while (search_queue.length) {
    let person = search_queue.shift();
    // Only search this person if you haven't already searched them
    if (searched.indexOf(person) === -1) {
      if (person_is_seller(person)) {
        console.log(person + ' is a mango seller!');
        return true;
      } else {
        search_queue = search_queue.concat(graph[person]);
        // Marks this person as searched
        searched.push(person);
      }
    }
  }
  return false;
}


search('you'); // thom is a mango seller!

File: Python_Books/算法图解/grokking_algorithms_COPY/06_breadth-first_search/lua/01_breadth-first_search.lua
-- Custom deque module
require "deque"

local function person_is_seller(name)
    return string.char(string.byte(name, -1)) == "m"
end

local graph = {}
graph["you"] = {"alice", "bob", "claire"}
graph["bob"] = {"anuj", "peggy"}
graph["alice"] = {"peggy"}
graph["claire"] = {"thom", "jonny"}
graph["anuj"] = {}
graph["peggy"] = {}
graph["thom"] = {}
graph["jonny"] = {}

local function search(name)
    local search_queue = deque:new()
    for _, value in pairs(graph[name]) do
        search_queue:push_right(value)
    end
    -- This array is how you keep track of which people you've searched before.
    local searched = {}
    while search_queue:len() > 0 do
        local person = search_queue:pop_left()
        -- Only search this person if you haven't already searched them.
        if not searched[person] then
            if person_is_seller(person) then
                print(person .. " is a mango seller!")
                return true
            else
                for _, value in pairs(graph[person]) do
                    search_queue:push_right(value)
                end
                -- Marks this person as searched
                searched[person] = true
            end
        end
    end
    return false
end

search("you")